gui-js/apps/minsky-electron/src/app/backend/backend.ts

```diff
@@ -0,0 +1,120 @@
+if (!Utility.isDevelopmentMode()) { //clobber logging in production
+  log.info=function(...args: any[]){};
+};
+
+/** returns true if RESTService call is logged in development mode */
+function logFilter(c: string) {
+  const logFilter=["mouseMove$", "requestRedraw$"];
+  for (var i in logFilter)
+    if (c.match(logFilter[i])) return false;
+  return true;
+}
+
+/** REST Service addon */
+export var restService = null;
+try {
+  restService = require('bindings')(join(addonDir, 'minskyRESTService.node'));
+} catch (error) {
+  log.error(error);
+}
+
+var errorDialog=(message: string)=>{};
+/** enable popup display of exception thrown by backend */
+export function setErrorDialog(f: (message: string)=>void) {errorDialog=f}
+
+/** core function to call into C++ object heirarachy */
+export function backend(command: string, ...args: any[]) {
+  if (!command) {
+    log.error('backend called without any command');
+    return {};
+  }
+  if (!restService) {
+    log.error('Rest Service not ready');
+    return {};
+  }
+  try {
+    var arg='';
+    if (args.length>0)
+      arg=JSON5.stringify(args);
+    const response = restService.call(command, arg);
+    if (logFilter(command))
+      log.info('Rest API: ',command,arg,"=>",response);
+    return JSON5.parse(response);
+  } catch (error) {
+    log.error('Rest API: ',command,arg,'=>Exception caught: ' + error?.message);
+    if (!dialog) throw error; // rethrow to force error in jest environment
+    if (command === '/minsky/canvas/item/importFromCSV') {
+      return importCSVerrorMessage;
+    } else {
+      if (error?.message) errorDialog(error.message);
+      return error?.message;
+    }
+  }
+}
+
+export class CppClass
+{
+  protected prefix: string;
+  constructor(prefix: string) {this.prefix=prefix;}
+  protected callMethod(method: string,...args)
+  {
+    return backend(this.prefix+"/"+method, ...args);
+  }
+  public properties(...args) {return backend(this.prefix, ...args);}
+};
+
+
+export class Pair<Key, Value extends CppClass> {
+  first: Key;
+  second: Value;
+  constructor(key: Key, value: Value) {
+    this.first=key;
+    this.second=value;
+  }
+};
+
+export class Map<Key, Value extends CppClass> extends CppClass
+{
+  valueType: any;  // stash a reference to the actual type here, for use in a new expression
+  constructor(prefix: string, valueType: any=null) {super(prefix); this.valueType=valueType;}
+  elem(key: Key) {
+    const cmd=this.prefix+"/@elem/"+JSON5.stringify(key)+"/second";
+    // if proxy type provided, instantiate that, otherwise return the current value
+    return new Pair<Key,Value>
+      (key,this.valueType? new this.valueType(cmd): this.callMethod(cmd));
+  }
+  insert(keyValue: Pair<Key, Value>) {this.callMethod("@insert",keyValue);}
+  erase(key: Key) {this.callMethod("@erase",key);}
+};
+
+export class Container<Key,Value=Key> extends CppClass
+{
+  type: any;
+  constructor(prefix: string, type: any=null) {super(prefix); this.type=type}
```

Comment:
Single-line formatting like this is kinda frowned upon in most circles I go... You're packing a lot of information in a small number of lines... which I like on some level, but I admit I had to read it a few times to fully grasp what's going on.

You can shorten the this.type = type stuff by declaring type as "protected type?" in the constructor. The question mark will also handle the optionality with null default (although I think it will use "undefined" instead of "null", which may have subtly different behavior!).

Depending on tsconfig settings, specifying the type of a variable as "any" may also not be strictly necessary.

Reply:
I'm used to the more condensed style. Vertical space is at a premium, you want maximum context in your editor window. So I find the more spaced out code styles slightly irritating, actually. However, I can happily work in code bases that insist on that sort of formatting if there's an automatic formatting tool that can be used - eg clang-format for C++ as used in another project I work with.

Reply:
Sounds good. The main thing I want is to avoid spending time in code reviews over "mere" style things, when it's more important to focus on maintainability, possible bugs and sometimes performance things. I've been on teams that had rigid style guides, and others that were a free-for-all. To be honest, consistency in style is overrated, but it is good to use idiomatic expressions where established.

Reply:
Style consistency can be useful for avoiding merging complications, though, especially when enforced automatically.
Also if it needed to be said: the reason I'm focusing on style related factors is that everything looks excellent on the other fronts. I don't expect my future reviews will have to deal with it much.